synedrion/src/tools/secret.rs

```diff
@@ -39,22 +73,349 @@ impl<T: Zeroize> Debug for Secret<T> {
-    type Target = SecretBox<T>;
+impl<T: Zeroize + for<'a> SubAssign<&'a T>> Sub<Secret<T>> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn sub(mut self, other: Secret<T>) -> Self::Output {
+        self -= &other;
+        self
+    }
+}
+
+// Multiplication
+
+impl<'a, T: Zeroize + MulAssign<&'a T>> MulAssign<&'a T> for Secret<T> {
+    fn mul_assign(&mut self, other: &'a T) {
+        self.expose_secret_mut().mul_assign(other)
+    }
+}
+
+impl<'a, T: Zeroize + MulAssign<&'a T>> MulAssign<&'a Secret<T>> for Secret<T> {
+    fn mul_assign(&mut self, other: &'a Secret<T>) {
+        self.mul_assign(other.expose_secret())
+    }
+}
+
+impl<'a, T: Zeroize + MulAssign<&'a T>> Mul<&'a T> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn mul(mut self, other: &'a T) -> Self::Output {
+        self *= other;
+        self
+    }
+}
+
+impl<T: Zeroize + Clone + for<'a> MulAssign<&'a T>> Mul<T> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn mul(mut self, other: T) -> Self::Output {
+        self *= &other;
+        self
+    }
+}
+
+impl<T: Zeroize + Clone + for<'a> MulAssign<&'a T>> Mul<T> for &Secret<T> {
+    type Output = Secret<T>;
+
+    fn mul(self, other: T) -> Self::Output {
+        let mut result: Secret<T> = self.clone();
+        result *= &other;
+        result
+    }
+}
+
+impl<T: Zeroize + for<'a> MulAssign<&'a T>> Mul<Secret<T>> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn mul(mut self, other: Secret<T>) -> Self::Output {
+        self *= &other;
+        self
+    }
+}
+
+impl<'a, T: Zeroize + MulAssign<&'a T>> Mul<&'a Secret<T>> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn mul(mut self, other: &'a Secret<T>) -> Self::Output {
+        self *= other.expose_secret();
+        self
+    }
+}
+
+// Division
+
+impl<'a, T: Zeroize + DivAssign<&'a NonZero<T>>> DivAssign<&'a NonZero<T>> for Secret<T> {
+    fn div_assign(&mut self, other: &'a NonZero<T>) {
+        self.expose_secret_mut().div_assign(other)
+    }
+}
+
+impl<T: Zeroize + for<'a> DivAssign<&'a NonZero<T>>> Div<NonZero<T>> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn div(mut self, other: NonZero<T>) -> Self::Output {
+        self /= &other;
+        self
+    }
+}
+
+// Remainder
+
+impl<'a, T: Zeroize + Clone + RemAssign<&'a NonZero<T>>> RemAssign<&'a NonZero<T>> for Secret<T> {
+    fn rem_assign(&mut self, other: &'a NonZero<T>) {
+        self.expose_secret_mut().rem_assign(other)
+    }
+}
+
+impl<'a, T: Zeroize + Clone + RemAssign<&'a NonZero<T>>> Rem<&'a NonZero<T>> for &Secret<T> {
+    type Output = Secret<T>;
+
+    fn rem(self, other: &'a NonZero<T>) -> Self::Output {
+        let mut result = self.clone();
+        result %= other;
+        result
+    }
+}
+
+// Summation
+
+impl<T: Zeroize + Clone + for<'a> AddAssign<&'a T> + Default> core::iter::Sum for Secret<T> {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.reduce(Add::add).unwrap_or(Secret::init_with(|| T::default()))
```

Comment:
Maybe we can impl `Default` for `Secret<T: Default>`?

Reply:
Yeah, that's a good idea

Reply:
Added